dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,455 @@
+            sbStore.append(visualizeStore(elseStore));
+        }
+        sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+        return sbStore.toString();
+    }
+
+    /**
+     * Helper method to simplify visualizing a special Block; it is useful when implementing a
+     * custom CFGVisualizer.
+     *
+     * @param sbb the special {@link Block}
+     * @param separator the separator String to use
+     * @return the String representation of the special {@link Block}
+     */
+    protected String visualizeSpecialBlockHelper(SpecialBlock sbb, String separator) {
+        String specialBlock = "";
+        switch (sbb.getSpecialType()) {
+            case ENTRY:
+                specialBlock = "<entry>" + separator;
+                break;
+            case EXIT:
+                specialBlock = "<exit>" + separator;
+                break;
+            case EXCEPTIONAL_EXIT:
+                specialBlock = "<exceptional-exit>" + separator;
+                break;
+        }
+        return specialBlock;
+    }
+
+    /**
+     * Helper method called by {@link #visualizeBlockHelper}. If possible, get the last {@link Node}
+     * of a {@link Block}.
+     *
+     * @param bb the {@link Block}
+     * @return the last {@link Node} or {@code null}
+     */
+    protected Node getLastNode(Block bb) {
+        Node lastNode;
+        switch (bb.getType()) {
+            case REGULAR_BLOCK:
+                List<Node> blockContents = ((RegularBlock) bb).getContents();
+                lastNode = blockContents.get(blockContents.size() - 1);
+                break;
+            case EXCEPTION_BLOCK:
+                lastNode = ((ExceptionBlock) bb).getNode();
+                break;
+            default:
+                lastNode = null;
+        }
+        return lastNode;
+    }
+
+    /**
+     * Generate the order of processing {@link Block}s.
+     *
+     * @param cfg the current control flow graph
+     * @return the IdentityHashMap which maps from {@link Block}s to their orders
+     */
+    protected IdentityHashMap<Block, List<Integer>> getProcessOrder(ControlFlowGraph cfg) {
+        IdentityHashMap<Block, List<Integer>> depthFirstOrder = new IdentityHashMap<>();
+        int count = 1;
+        for (Block b : cfg.getDepthFirstOrderedBlocks()) {
+            depthFirstOrder.computeIfAbsent(b, k -> new ArrayList<>());
+            depthFirstOrder.get(b).add(count++);
+        }
+        return depthFirstOrder;
+    }
+
+    @Override
+    public String visualizeStore(S store) {
+        return store.visualize(this);
+    }
+
+    /**
+     * Return the header of the generated graph. Called by {@link
+     * #generateGraphHelper(ControlFlowGraph, Block, Analysis)}.
+     *
+     * <p>This abstract method needs to be implemented to customize the output.
+     *
+     * @return the String representation of the header of the control flow graph
+     */
+    protected abstract String visualizeGraphHeader();
```

Comment:
How about moving all abstract methods together? It isn't quite clear why some are earlier and some are down here.